### ISSUE TITLE
Lumen/Laravel para criação de API em sistema com outro Framework

### ISSUE THREAD

**Issue by zabaala** [laravelbrasil/forum#55]
Vale a pena utilizar o Laravel ou Lumen para criar uma API que consuma uma base de dados de uma outra aplicação e gerenciada por um outro Framework?

## Problema Encontrado
Estou trabalhando em cima de uma aplicação criada em Yii Framework. Ela funciona! Mas o Yii não é um framework muito bacana, convenhamos, principalmente na versão 1.x. Então considero ruim continuar criando novas implementações no sistema atual, sem contar no quão oneroso é executar essas tarefas em cima dele.  No caso, a nível de informação, o projeto foi desenvolvido por um outro programador e eu só estou dando continuidade. 

---

## Abordagem
No momento não posso abandonar o sistema atual desenvolver um novo, por motivos que sequer valem a pena serem citados. Então pensei em manter o sistema lá e começar a migra-lo pra uma API, usando um outro framework; criar uma outra camada pra consumo; Não sei... Procuro por algo que me deixe mais livre, mais produtivo e que me permita criar algo com a qualidade que o framework atual não me permite. 

Pensei em criar algo com Lumen ou Laravel. Mas ainda não tenho nada definido, por isso rogo por sua ajuda, caro companheiro de difíceis missões. 😂 

Gostaria de saber se esse é um pensamento ruim; se existem alternativas melhores. Enfim... preciso muito de luz, pq esse Yii tem tirado a minha paciência.

**Comment by cagartner** [comment]
Acredito ser uma ótima abordagem, eu pessoalmente já usei o Laravel em conjunto com outros projetos pelo prazo e minha familiariadade com a ferramenta.

Eu pessoalmente prefiro o Laravel mesmo para API's por ser mais completo, não gosto muito do Lumen, no final você vai acabar adicionando tanta coisa nele que vai ficar igual ao Laravel rsrsrs.

Tudo depende do tamanho e necessidade do serviço que vai fazer.

Se for simples apenas para listar alguma coisa do banco e vai consultar no máximo 5 tabelas, usa o lumen, se quer algo mais completo com envio de e-mail, jobs e tal, usa o Laravel.

Espero que ajudado!

**Comment by zabaala** [comment]
@cagartner Pois é. 

Me sinto a vontade em fazer algo desse tipo, mas o que o que me angustia é não conseguir visualizar os problemas que terei por tomar essa decisão.

Minha ideia é partir pra uma transição mesmo. Num processo paulatino, esperando ter algo bem consistente destro de 6... 12 meses. Mas queria ver um pouco dos "contras" dessa iniciativa.

**Comment by cagartner** [comment]
@zabaala, segue uns problemas que posso enxergar a ocorrer.

1. Um sistema ficar fora ou ocioso e outro não, no caso você vai precisar dar manutensão em dois sistemas e não 1
2. Se tiver mais gente na equipe, pode ocorrer de alguém não tiver o conhecimento em Laravel e a curva de aprendizado ser um problema para o Gerente e demais pessoas superiores
3. Sistemas de autenticação, no yii pode ser mais demorado implementar um sistema de autenticação atual, ou você terá que colocar uma autenticação ultrapassada no Laravel.

De resto não consigo visualizar mais pontos de problemas, como no seu caso é uma implementação **Nova**  não afetaria tanto o sistema antigo, mas caso seja uma refatoração do que já existe um problema pode ser o tempo que você irá levar para refatorar e também lembrar de não deixar alguma regra de negócio existente no Yii fora do Laravel.

Sempre que rola refatoração a atenção deve ser quadriplicada rsrsrs.

Vamos aguardar novas sugestões e cenários bons e ruins da comunidade :)

**Comment by flyingluscas** [comment]
@zabaala Recentemente eu desenvolvi uma API bem simples usando Laravel, depois de um tempo achei melhor migrar para Lumen, mas acabei na mesma situação que o @cagartner citou a cima, tive que adicionar algumas features no Lumen manualmente, e no final das contas ficou igual ao Laravel, mas com um pouco mais de trabalho.

**Comment by renanoliveira0** [comment]
Isso não seria uma refatoração para microservices?

**Comment by meneguite** [comment]
Olha já passei por situação parecida... E sua opção não é ruim mas tem alguns pontos de atenção.. por exemplo o que o @cagartner comentou.. para adicionar uma pessoa ao grupo ele teria que ter conhecimento em dois frameworks que são relativamente diferentes..
Uma outra possibilidade que vejo seria criar pequenos pacotes redistribuiveis e independentes de framework que poderia em um momento futuro reaproveitar em uma possível migração para o Laravel/Lumen..

**Comment by dorianneto** [comment]
Concordo com o @renanoliveira0, isso cheira a refatoração de monolítico para microservices. Acredito que o foco inicial deva ser a discussão da rearquitetura do software e só depois a escolha do framework, caso contrário, tu vai sofrer bastante no futuro com os pontos levantados pelo @cagartner.

Tem um artigo muito bom que sugere como deve ser essa refatoração: https://www.nginx.com/blog/refactoring-a-monolith-into-microservices/

Esse vídeo também pode te ajudar nessa decisão: https://www.youtube.com/watch?v=uJ8zBJdT6Is

**Comment by zabaala** [comment]
@dorianneto É bem isso mesmo que estou pensando e exatamente por isso que fiz essa pergunta, pra tentar manifestar opniões como a sua e as dos demais, pq queria validar um pouco a minha idéia.
